osu.Game/Screens/Play/SubmittingPlayer.cs

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+using osu.Framework.Allocation;
+using osu.Framework.Logging;
+using osu.Framework.Screens;
+using osu.Game.Online.API;
+using osu.Game.Online.Rooms;
+using osu.Game.Scoring;
+
+namespace osu.Game.Screens.Play
+{
+    /// <summary>
+    /// A player instance which supports submitting scores to an online store.
+    /// </summary>
+    public abstract class SubmittingPlayer : Player
+    {
+        /// <summary>
+        /// The token to be used for the current submission. This is fetched via a request created by <see cref="CreateTokenRequest"/>.
+        /// </summary>
+        private long? token;
+
+        [Resolved]
+        private IAPIProvider api { get; set; }
+
+        protected SubmittingPlayer(PlayerConfiguration configuration = null)
+            : base(configuration)
+        {
+        }
+
+        protected override void LoadAsyncComplete()
+        {
+            // Token request construction should happen post-load to allow derived classes to potentially prepare DI backings that are used to create the request.
+            var tcs = new TaskCompletionSource<bool>();
+
+            if (!api.IsLoggedIn)
+            {
+                handleTokenFailure(new InvalidOperationException("API is not online."));
+                return;
```

Comment:
This method kinda bugs me a bit, I suppose it's fine if you really don't want to change it, but these `return;`s look a bit suspicious because they're done before calling `base.LoadAsyncComplete()`.

Would you have anything against extracting all this to a new method, something like:
```
protected override void LoadAsyncComplete()
{
    if (!prepareToken())
        return; // If it really needs to be done this way, otherwise just return void and call it after base.LoadAsyncComplete()?
    base.LoadAsyncComplete();
}

private bool prepareToken()
{
    if (cond)
        return handleTokenFailure(...);

    ...

    return true;

    bool handleTokenFailure()
    {
        if (HandleTokenRetrievalFailure(exception))
        {
            ...
        }

        return false;
    }
}

```

Reply:
Yeah, also found this hard to read so i think that will work better.

The server side changes are available at https://github.com/ppy/osu-web/pull/7371 but I haven't tested against it yet.